File: Data Structures & Algorithms/combination-target-sum/submission-3.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        seen = set()
        res = []
        def backtrack(curT, l):
            nonlocal res
            if curT < 0:
                return
            for i in range(len(nums) -1, -1, -1):
                if nums[i] > curT:
                    continue
                goofy = l + [nums[i]] 
                goofy.sort()
                if nums[i] == curT and goofy not in res:
                    res.append(goofy)
                elif nums[i] < target and tuple(goofy) not in seen:
                    seen.add(tuple(goofy))
                    backtrack(curT - nums[i], goofy)
                    
        backtrack(target, [])
        return res

File: Data Structures & Algorithms/subsets/submission-3.cpp
class Solution {
public:
    void backtrack(int i, vector<vector<int>> &r, vector<int> &s, vector<int> &n) {
            if (i >= n.size()) {
                r.push_back(s);
                return;
            }
            s.push_back(n[i]);
            backtrack(i + 1, r, s, n);
            s.pop_back();
            backtrack(i + 1, r, s, n);
        }  
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> sub;
        backtrack(0, res, sub, nums);
        return res;
    }
};
